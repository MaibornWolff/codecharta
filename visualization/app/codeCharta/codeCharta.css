*, *:before, *:after {
    -webkit-box-sizing: border-box;
    -moz-box-sizing: border-box;
    box-sizing: border-box;
}

html, body {
    -webkit-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    user-select: none;
    width: 100%;
    height: 100%;
    margin: 0;
    padding: 0;
    overflow: hidden;
    font-size: 1em;
    cursor: default;
    color: #5A585A;
    font-family: Roboto, sans-serif;
    font-weight: 300;
}

#codeMap {
    position: absolute;
    z-index: 0;
    top: 50px;
    left: 0;
    width: 100%;
    height: 100%;
}

#hud {
    position: absolute;
    z-index: 10;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    max-width: none !important;
    pointer-events: none;
    padding-left:10px;
    padding-right:10px;
}

revision-chooser-component, legend-panel-component, file-chooser-panel-directive, settings-panel-component, .mwlink {
    pointer-events: auto;
}

#title {
    text-transform: uppercase;
    font-size: 0.9em;
    line-height: 1.4em;
    float: right;
    margin: 0.5em 1.6em 0 0;
    color: #5A585A;
    font-family: Roboto, sans-serif;
    font-weight: 300;
}

#title:first-child {
    font-size: 1.82em;
}

#title:last-child {
    font-size: 1.5em;
}

.collapsible > li {
    background-color: #ffffff !important;
}

.col-backdrop {
    background-color: rgba(0,0,0,0.1);
}

.bottom-row {
    position: fixed;
    bottom: 0;
    width:100%;
}

.mwlink > img{
    width: 100%;
    max-width: 4em;
    float:right;
}

.chip {
    background-color: #efefef;
}

.row, .col {
    min-height: 0px !important;
}


/*

#actionButtons {
    position: fixed;
    bottom: 2em;
    right: 2em;
    z-index: 1000;
}

#settingsPanel {
    position: fixed;
    right: 0;
    width: 0px;
    height: 100%;
    background: #fff;
    z-index: 1000;
    box-shadow: 0 0 10px #000;
}

#detailWindow {
    position: fixed;
    z-index: 999;
    bottom: 0.5em;
    left: 0.5em;
    color:#5A585A;
    font-size: 0.8em;
}

#legendButton {
    position: absolute;
    top: 11.5em;
    left: -2.9em;
    -webkit-transform: rotate(90deg);
    -moz-transform: rotate(90deg);
    -o-transform: rotate(90deg);
    -ms-transform: rotate(90deg);
    transform: rotate(90deg);
}

#legendPanel {
    position: fixed;
    top: 9em;
    left: -500px;
}

#revisionButton {
    position: absolute;
    top: 21em;
    left: -3.5em;
    -webkit-transform: rotate(90deg);
    -moz-transform: rotate(90deg);
    -o-transform: rotate(90deg);
    -ms-transform: rotate(90deg);
    transform: rotate(90deg);
}

#mapButton {
    position: absolute;
    top: 21em;
    left: -2.3em;
    -webkit-transform: rotate(90deg);
    -moz-transform: rotate(90deg);
    -o-transform: rotate(90deg);
    -ms-transform: rotate(90deg);
    transform: rotate(90deg);
}

#revisionChooser {
    position: fixed;
    top: 18.5em;
    left: -500px;
}

.rzslider .rz-pointer, .rzslider .rz-selection {
    background: #26a69a !important;
}

.rzslider .rz-bar {
    background: rgb(196,196,196);
}*/