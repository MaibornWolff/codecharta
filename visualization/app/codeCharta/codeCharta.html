<div id="hud" class="container">

    <div class="row">

        <div class="col l3 m6 s12 col-backdrop">
            <legend-panel-component></legend-panel-component>
        </div>

        <div class="col l3 m6 s12 col-backdrop">
            <revision-chooser-component></revision-chooser-component>
        </div>

        <div class="col l3 m6 s12 col-backdrop">
            <file-chooser-panel-directive></file-chooser-panel-directive>
        </div>

        <div class="col l3 m6 s12 col-backdrop">
            <settings-panel-component></settings-panel-component>
        </div>

    </div>

    <div class="row bottom-row">

        <div class="col l3 m6 s12 col-backdrop">
            <detail-panel-component></detail-panel-component>
        </div>

    </div>

    <div id="titlebox">
        <a class="mwlink" href="https://www.maibornwolff.de" target="_blank"><img src="./logo.png" alt="MaibornWolff"></a>
        <h1 id="title"><span>CodeCharta {{$ctrl.pkg.version}}</span></h1>
    </div>

</div>

<code-map-component></code-map-component>