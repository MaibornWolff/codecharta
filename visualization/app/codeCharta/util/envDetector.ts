export function isStandalone(): boolean {
	return JSON.parse(process.env.STANDALONE)
}

export function isDevelopment(): boolean {
	return process.env.DEV === undefined ? false : JSON.parse(process.env.DEV);
}
