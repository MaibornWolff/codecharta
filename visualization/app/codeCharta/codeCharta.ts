"use strict";

import "./core/core.module";
import "./ui/ui";

import {codeChartaComponent, CodeChartaController} from "./codeCharta.component";

import angular from "angular";

angular.module(
    "app.codeCharta",
    ["app.codeCharta.core", "app.codeCharta.ui"]
);

angular.module("app.codeCharta").component(
    codeChartaComponent.selector,
    codeChartaComponent
);


