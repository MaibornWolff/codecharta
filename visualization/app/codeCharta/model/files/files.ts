import { CCFile } from "../../codeCharta.model"

export interface FileState {
	file: CCFile
	selectedAs: FileSelectionState
}

export enum FileSelectionState {
	Single = "Single",
	Reference = "Reference",
	Comparison = "Comparison",
	Partial = "Partial",
	None = "None"
}
