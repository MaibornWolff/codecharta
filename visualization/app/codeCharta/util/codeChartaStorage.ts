"use strict"

import fs from "fs"
import { isStandalone } from "./envDetector"

export class CodeChartaStorage implements Storage {
	private storageEngine: Storage = isStandalone() ? new CodeChartaFileStorageEngine() : localStorage
	readonly length = -1

	clear(): void {
		this.storageEngine.clear()
	}

	getItem(key: string): string | null {
		return this.storageEngine.getItem(key)
	}

	key(index: number): string | null {
		return this.storageEngine.key(index)
	}

	removeItem(key: string): void {
		this.storageEngine.removeItem(key)
	}

	setItem(key: string, value: string): void {
		this.storageEngine.setItem(key, value)
	}
}

export class CodeChartaFileStorageEngine implements Storage {
	private fileStoragePath = `${fs.realpathSync(".")}/app/codeCharta/ressources/fileStorage/`

	// TODO: Find out, how to implement this.
	readonly length = -1

	clear(): void {
		fs.readdirSync(this.fileStoragePath).forEach(file => {
			fs.unlinkSync(this.fileStoragePath + file)
		})
	}

	getItem(key: string): string | null {
		return fs.readFileSync(this.fileStoragePath + key).toString()
	}

	key(index: number): string | null {
		return fs.readdirSync(this.fileStoragePath)[index] ?? null
	}

	removeItem(key: string): void {
		fs.unlinkSync(this.fileStoragePath + key)
	}

	setItem(key: string, value: string): void {
		fs.writeFileSync(this.fileStoragePath + key, value, { encoding: "utf-8" })
	}
}
