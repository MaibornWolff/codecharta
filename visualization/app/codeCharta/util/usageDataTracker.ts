"use strict"

import { getVisibleFileStates, isPartialState } from "../model/files/files.helper"
import { CodeMapNode, NodeType } from "../codeCharta.model"
import { isActionOfType } from "./reduxHelper"
import { AreaMetricActions } from "../state/store/dynamicSettings/areaMetric/areaMetric.actions"
import { HeightMetricActions } from "../state/store/dynamicSettings/heightMetric/heightMetric.actions"
import { ColorMetricActions } from "../state/store/dynamicSettings/colorMetric/colorMetric.actions"
import { BlacklistActions } from "../state/store/fileSettings/blacklist/blacklist.actions"
import { FocusedNodePathActions } from "../state/store/dynamicSettings/focusedNodePath/focusedNodePath.actions"
import md5 from "md5"
import { APIVersions } from "../codeCharta.api.model"
import { getAsApiVersion } from "./fileValidator"
import { hierarchy } from "d3-hierarchy"
import { getMedian, pushSorted } from "./nodeDecorator"
import { ColorRangeActions } from "../state/store/dynamicSettings/colorRange/colorRange.actions"
import { FileState } from "../model/files/files"

interface MetaDataTrackingItem {
	mapId: string
	codeChartaApiVersion: string
	creationTime: number
	exportedFileSizeInBytes: number
	statisticsPerLanguage: StatisticsPerLanguage
	repoCreationDate: string
}

interface StatisticsPerLanguage {
	[languageName: string]: LanguageStatistics
}

interface LanguageStatistics {
	numberOfFiles: number
	maxFilePathDepth: number
	avgFilePathDepth: number
	metrics: {
		[metricName: string]: MetricStatistics
	}
}

interface MetricStatistics {
	min: number
	firstQuantile: number
	median: number
	thirdQuantile: number
	max: number
	outliers: number[]
	variance: number
	standardDeviation: number
	variationCoefficient: number
	totalSum: number
	numberOfFiles: number
	avg: number
	metricValues: number[]
}

export const TRACKING_DATA_LOCAL_STORAGE_ELEMENT = "CodeCharta::usageData"

function isTrackingAllowed(files: FileState[]) {
	const singleFileStates = getVisibleFileStates(files)

	// replaced isSingleState with isPartialState, workaround so tests still work, rebuild is planned in a future story
	if (!isPartialState(files) || singleFileStates.length > 1) {
		return false
	}

	const fileApiVersion = getAsApiVersion(singleFileStates[0].file.fileMeta.apiVersion)
	const supportedApiVersion = getAsApiVersion(APIVersions.ONE_POINT_ZERO)

	return (
		fileApiVersion.major > supportedApiVersion.major ||
		(fileApiVersion.major === supportedApiVersion.major && fileApiVersion.minor >= supportedApiVersion.minor)
	)
}

export function trackMapMetaData(files: FileState[]) {
	if (!isTrackingAllowed(files)) {
		return
	}

	const singleFileStates = getVisibleFileStates(files)
	const fileNodes: CodeMapNode[] = getFileNodes(singleFileStates[0].file.map)
	const fileMeta = singleFileStates[0].file.fileMeta

	const trackingDataItem: MetaDataTrackingItem = {
		mapId: fileMeta.fileChecksum,
		codeChartaApiVersion: fileMeta.apiVersion,
		creationTime: Date.now(),
		exportedFileSizeInBytes: fileMeta.exportedFileSize,
		statisticsPerLanguage: mapStatisticsPerLanguage(fileNodes),
		repoCreationDate: fileMeta.repoCreationDate
	}

	// Make sure that only files within usageData can be read
	const fileChecksum = trackingDataItem.mapId.replace(/\//g, "")

	try {
		localStorage.setItem(`${TRACKING_DATA_LOCAL_STORAGE_ELEMENT}/${fileChecksum}-meta`, JSON.stringify(trackingDataItem))
	} catch {
		// ignore it
	}
}

function getFileNodes(node: CodeMapNode): CodeMapNode[] {
	const fileNodes: CodeMapNode[] = []
	for (const { data } of hierarchy(node)) {
		if (data.type === NodeType.FILE) {
			fileNodes.push(data)
		}
	}
	return fileNodes
}

function mapStatisticsPerLanguage(fileNodes: CodeMapNode[]): StatisticsPerLanguage {
	const statisticsPerLanguage: StatisticsPerLanguage = {}

	const sumOfFilePathDepths: { [languageName: string]: number } = {}
	const metricValues: { [languageName: string]: { [metricName: string]: number[] } } = {}
	const unsortedMetricValues: { [languageName: string]: { [metricName: string]: number[] } } = {}

	for (const fileNode of fileNodes) {
		const fileLanguage = getFileExtension(fileNode.name)
		if (fileLanguage === null) {
			continue
		}

		//TODO: How to initialize automatically?
		if (metricValues[fileLanguage] === undefined) {
			metricValues[fileLanguage] = {}
		}
		if (unsortedMetricValues[fileLanguage] === undefined) {
			unsortedMetricValues[fileLanguage] = {}
		}
		if (sumOfFilePathDepths[fileLanguage] === undefined) {
			sumOfFilePathDepths[fileLanguage] = 0
		}

		// Initialize statistics object for unseen metric of language
		if (!Object.prototype.hasOwnProperty.call(statisticsPerLanguage, fileLanguage)) {
			initializeLanguageStatistics(fileNode, statisticsPerLanguage, fileLanguage)
		}

		const currentLanguageStats = statisticsPerLanguage[fileLanguage]
		currentLanguageStats.numberOfFiles += 1

		const currentPathDepth = getPathDepth(fileNode.path)
		sumOfFilePathDepths[fileLanguage] += currentPathDepth
		currentLanguageStats.maxFilePathDepth = Math.max(currentLanguageStats.maxFilePathDepth, currentPathDepth)
		currentLanguageStats.avgFilePathDepth = sumOfFilePathDepths[fileLanguage] / currentLanguageStats.numberOfFiles

		for (const metricName of Object.keys(fileNode.attributes)) {
			const metricStatistics: MetricStatistics = currentLanguageStats.metrics[metricName]
			if (metricStatistics === undefined) {
				// Skip if file has different set of metrics compared to other files of the same language
				continue
			}

			if (metricValues[fileLanguage][metricName] === undefined) {
				metricValues[fileLanguage][metricName] = []
			}
			if (unsortedMetricValues[fileLanguage][metricName] === undefined) {
				unsortedMetricValues[fileLanguage][metricName] = []
			}

			const valuesOfMetric = metricValues[fileLanguage][metricName]
			const unsortedValuesOfMetric = unsortedMetricValues[fileLanguage][metricName]
			const currentMetricValue = fileNode.attributes[metricName]
			pushSorted(valuesOfMetric, currentMetricValue)
			unsortedValuesOfMetric.push(currentMetricValue)

			metricStatistics.median = getMedian(valuesOfMetric)
			metricStatistics.max = Math.max(metricStatistics.max, currentMetricValue)
			metricStatistics.min = Math.min(metricStatistics.min, currentMetricValue)
			metricStatistics.numberOfFiles += 1
			metricStatistics.totalSum += currentMetricValue
			metricStatistics.avg = metricStatistics.totalSum / metricStatistics.numberOfFiles

			metricStatistics.variance = getVariance(valuesOfMetric)
			metricStatistics.standardDeviation = Math.sqrt(metricStatistics.variance)
			metricStatistics.variationCoefficient = metricStatistics.standardDeviation / metricStatistics.avg

			metricStatistics.metricValues = unsortedValuesOfMetric

			let valuesFirstHalf: number[]
			let valuesSecondHalf: number[]

			if (valuesOfMetric.length % 2 === 0) {
				valuesFirstHalf = valuesOfMetric.slice(0, valuesOfMetric.length / 2)
				valuesSecondHalf = valuesOfMetric.slice(valuesOfMetric.length / 2, valuesOfMetric.length)
			} else {
				valuesFirstHalf = valuesOfMetric.slice(0, valuesOfMetric.length / 2)
				valuesSecondHalf = valuesOfMetric.slice(valuesOfMetric.length / 2 + 1, valuesOfMetric.length)
			}

			metricStatistics.firstQuantile = getMedian(valuesFirstHalf)
			metricStatistics.thirdQuantile = getMedian(valuesSecondHalf)

			const interQuartileRange = metricStatistics.thirdQuantile - metricStatistics.firstQuantile
			const upperOutlierBound = metricStatistics.thirdQuantile + 1.5 * interQuartileRange

			metricStatistics.outliers = valuesOfMetric.filter(function (value) {
				return value > upperOutlierBound
			})
		}
	}

	return statisticsPerLanguage
}

function getVariance(array) {
	const n = array.length
	const mean = array.reduce((a, b) => a + b) / n
	return array.map(x => Math.pow(x - mean, 2)).reduce((a, b) => a + b) / n
}

function initializeLanguageStatistics(fileNode: CodeMapNode, statisticsPerLanguage: StatisticsPerLanguage, fileLanguage: string) {
	const initialPathDepth = getPathDepth(fileNode.path)
	const statisticsCurrentLanguage = (statisticsPerLanguage[fileLanguage] = {
		metrics: {},
		numberOfFiles: 0,
		maxFilePathDepth: initialPathDepth,
		avgFilePathDepth: initialPathDepth
	})

	for (const metricName of Object.keys(fileNode.attributes)) {
		statisticsCurrentLanguage.metrics[metricName] = {
			avg: 0,
			max: fileNode.attributes[metricName],
			median: 0,
			min: fileNode.attributes[metricName],
			numberOfFiles: 0,
			totalSum: 0,
			firstQuantile: 0,
			thirdQuantile: 0,
			variance: 0,
			standardDeviation: 0,
			variationCoefficient: 0,
			outliers: []
		}
	}
}

function getPathDepth(path: string): number {
	return path.split("/").length
}

function getFileExtension(filePath: string): string {
	const lastDotPosition = filePath.lastIndexOf(".")
	return lastDotPosition !== -1 ? filePath.slice(lastDotPosition + 1) : null
}

interface SettingChangedEventPayload {
	eventName: string
	newValue: unknown
}

interface NodeInteractionEventPayload {
	eventName: string
	id: string
	type?: string
	nodeType?: string
	attributes?: MetricStatistics
}

interface EventTrackingItem {
	mapId: string
	eventType: string
	eventTime: number
	payload: SettingChangedEventPayload | NodeInteractionEventPayload
}

export function trackEventUsageData(actionType: string, files: FileState[], payload?: any) {
	if (!isTrackingAllowed(files)) {
		return
	}

	const singleFileStates = getVisibleFileStates(files)
	const fileMeta = singleFileStates[0].file.fileMeta

	const eventTrackingItem = buildEventTrackingItem(fileMeta.fileChecksum, actionType, payload)
	if (eventTrackingItem === null) {
		return
	}

	// Make sure that only files within usageData can be read
	const fileChecksum = getVisibleFileStates(files)[0].file.fileMeta.fileChecksum.replace(/\//g, "")

	let appendedEvents = ""
	try {
		appendedEvents = localStorage.getItem(`${TRACKING_DATA_LOCAL_STORAGE_ELEMENT}/${fileChecksum}-events`)
	} catch {
		// ignore, it no events item exists
	}

	try {
		if (appendedEvents.length > 0) {
			appendedEvents += "\n"
		}
		localStorage.setItem(
			`${TRACKING_DATA_LOCAL_STORAGE_ELEMENT}/${fileChecksum}-events`,
			appendedEvents + JSON.stringify(eventTrackingItem)
		)
	} catch {
		// ignore tracking errors
	}
}

function buildEventTrackingItem(
	mapId: string,
	actionType: string,
	payload?: string & Record<string, string & MetricStatistics>
): EventTrackingItem | null {
	if (isSettingChangedEvent(actionType)) {
		return {
			mapId,
			eventType: "setting_changed",
			eventTime: Date.now(),
			payload: {
				eventName: actionType,
				newValue: payload
			}
		}
	}

	if (actionType === BlacklistActions.ADD_BLACKLIST_ITEM || actionType === BlacklistActions.REMOVE_BLACKLIST_ITEM) {
		return {
			mapId,
			eventType: "node_interaction",
			eventTime: Date.now(),
			payload: {
				eventName: actionType,
				id: md5(payload.path),
				type: payload.type,
				nodeType: payload.nodeType,
				attributes: payload.attributes
			}
		}
	}

	if (isActionOfType(actionType, FocusedNodePathActions) && payload !== "") {
		return {
			mapId,
			eventType: "node_interaction",
			eventTime: Date.now(),
			payload: {
				eventName: actionType,
				id: md5(payload)
			}
		}
	}

	return null
}

function isSettingChangedEvent(actionType: string) {
	return (
		isActionOfType(actionType, AreaMetricActions) ||
		isActionOfType(actionType, HeightMetricActions) ||
		isActionOfType(actionType, ColorMetricActions) ||
		isActionOfType(actionType, ColorRangeActions) ||
		actionType === BlacklistActions.SET_BLACKLIST
	)
}
