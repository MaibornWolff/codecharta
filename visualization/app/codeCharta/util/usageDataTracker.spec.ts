import { CCFile, CodeMapNode, NodeType, State } from "../codeCharta.model"
import { trackEventUsageData, trackMapMetaData } from "./usageDataTracker"
import * as EnvironmentDetector from "./envDetector"
import * as FilesHelper from "../model/files/files.helper"
import { FileState } from "../model/files/files"
import { APIVersions } from "../codeCharta.api.model"
import { CodeChartaStorage } from "./codeChartaStorage"
import { HeightMetricActions } from "../state/store/dynamicSettings/heightMetric/heightMetric.actions"
import { klona } from "klona"
jest.mock("./codeChartaStorage")

describe("UsageDataTracker", () => {
	// provide some default state properties
	const stateStub = {
		appSettings: {
			experimentalFeaturesEnabled: true,
			showMetricLabelNameValue: undefined,
			isWhiteBackground: false,
			camera: { x: 1, y: 2, z: 3 }
		},
		dynamicSettings: {},
		fileSettings: {}
	} as State

	let singleFileState: FileState
	beforeEach(() => {
		jest.resetAllMocks()

		singleFileState = {
			file: {
				fileMeta: {
					fileChecksum: "invalid-md5-sample-checksum",
					apiVersion: APIVersions.ONE_POINT_ONE,
					exportedFileSize: 999
				},
				map: {}
			} as CCFile
		} as FileState
	})

	function mockTrackingToBeAllowed() {
		jest.spyOn(EnvironmentDetector, "isStandalone").mockReturnValue(true)
		jest.spyOn(FilesHelper, "isSingleState").mockReturnValue(true)
		jest.spyOn(FilesHelper, "getVisibleFileStates").mockReturnValue([singleFileState])
	}

	describe("trackMetaUsageData", () => {
		const setItemMock = jest.fn()
		CodeChartaStorage.prototype.setItem = setItemMock

		afterAll(() => {
			jest.resetAllMocks()
			jest.unmock("./codeChartaStorage")
		})

		it("should not track in web version, in multi/delta mode or for more than one uploaded file", () => {
			jest.spyOn(EnvironmentDetector, "isStandalone").mockReturnValue(false)
			jest.spyOn(FilesHelper, "isSingleState").mockReturnValue(true)
			jest.spyOn(FilesHelper, "getVisibleFileStates").mockReturnValue([{} as FileState])

			trackMapMetaData(stateStub.files)

			jest.spyOn(EnvironmentDetector, "isStandalone").mockReturnValue(true)
			jest.spyOn(FilesHelper, "isSingleState").mockReturnValue(false)
			jest.spyOn(FilesHelper, "getVisibleFileStates").mockReturnValue([{} as FileState])

			trackMapMetaData(stateStub.files)

			jest.spyOn(EnvironmentDetector, "isStandalone").mockReturnValue(true)
			jest.spyOn(FilesHelper, "isSingleState").mockReturnValue(true)
			jest.spyOn(FilesHelper, "getVisibleFileStates").mockReturnValue([{} as FileState, {} as FileState])

			trackMapMetaData(stateStub.files)

			expect(setItemMock).not.toHaveBeenCalled()
		})

		it("should not track maps from old API versions", () => {
			singleFileState.file.fileMeta.apiVersion = APIVersions.ZERO_POINT_ONE
			mockTrackingToBeAllowed()
			trackMapMetaData(stateStub.files)

			singleFileState.file.fileMeta.apiVersion = "0.9"
			trackMapMetaData(stateStub.files)

			expect(setItemMock).not.toHaveBeenCalled()
		})

		it("should track from API version 1.0", () => {
			singleFileState.file.map = { path: "/root" } as CodeMapNode

			mockTrackingToBeAllowed()
			trackMapMetaData(stateStub.files)

			singleFileState.file.fileMeta.apiVersion = "1.0"
			trackMapMetaData(stateStub.files)

			singleFileState.file.fileMeta.apiVersion = "2.0"
			trackMapMetaData(stateStub.files)

			expect(setItemMock).toHaveBeenCalledTimes(3)
		})

		it("should track files with multiple programming languages properly", () => {
			singleFileState.file.map = mapStub

			assertMetaDataSnapshot()
		})

		it("should not consider files with divergent metrics for statistic calculation", () => {
			const mapStubDivergentMetrics = klona(mapStub)
			mapStubDivergentMetrics.children[1].attributes = { rloc: 0 }
			singleFileState.file.map = mapStubDivergentMetrics

			assertMetaDataSnapshot()
		})

		function assertMetaDataSnapshot() {
			mockTrackingToBeAllowed()
			jest.spyOn(Date, "now").mockReturnValue(1_612_369_999_999)

			const expectSetItemSnapshot = (CodeChartaStorage.prototype.setItem = jest.fn().mockImplementation((_, value) => {
				expect(value).toMatchSnapshot()
			}))

			trackMapMetaData(stateStub.files)

			expect(expectSetItemSnapshot).toHaveBeenCalledTimes(1)
		}
	})

	describe("trackEventUsageData", () => {
		afterAll(() => {
			jest.resetAllMocks()
			jest.unmock("./codeChartaStorage")
		})

		let getItemMock
		let expectSetItemSnapshot

		beforeEach(() => {
			jest.spyOn(EnvironmentDetector, "isStandalone").mockReturnValue(true)
			jest.spyOn(FilesHelper, "isSingleState").mockReturnValue(true)
			jest.spyOn(FilesHelper, "getVisibleFileStates").mockReturnValue([singleFileState])
			jest.spyOn(Date, "now").mockReturnValue(1_612_428_357_566)

			getItemMock = CodeChartaStorage.prototype.getItem = jest.fn().mockReturnValue("some-already-tracked-events-from-file-storage")
			expectSetItemSnapshot = CodeChartaStorage.prototype.setItem = jest.fn().mockImplementation((_, value) => {
				expect(value).toMatchSnapshot()
			})
		})

		function expectEventHasBeenTracked() {
			expect(FilesHelper.getVisibleFileStates).toHaveBeenCalledTimes(3)
			expect(getItemMock).toHaveBeenCalledWith("usageData/invalid-md5-sample-checksum-events")
			expect(expectSetItemSnapshot).toHaveBeenCalledTimes(1)
		}

		it("should not track when tracking is not allowed", () => {
			jest.spyOn(EnvironmentDetector, "isStandalone").mockReturnValue(false)

			trackEventUsageData(HeightMetricActions.SET_HEIGHT_METRIC, stateStub.files, "newHeightMetricValue")

			// A second call would indicate that the tracking has not been cancelled as expected
			expect(FilesHelper.getVisibleFileStates).toHaveBeenCalledTimes(1)
		})

		it("should track an event", () => {
			trackEventUsageData(HeightMetricActions.SET_HEIGHT_METRIC, stateStub.files, "newHeightMetricValue")
			expectEventHasBeenTracked()
		})
	})

	const mapStub: CodeMapNode = {
		name: "root",
		attributes: {},
		type: NodeType.FOLDER,
		isExcluded: false,
		isFlattened: false,
		children: [
			{
				name: "doNotTrackInvalidFileWithMissingExtension",
				path: "/root/doNotTrackInvalidFileWithMissingExtension",
				type: NodeType.FILE,
				attributes: { rloc: 100, functions: 10, mcc: 1 },
				isExcluded: false,
				isFlattened: false,
				link: "http://www.google.de"
			},
			{
				name: "bigLeaf.ts",
				path: "/root/bigLeaf.ts",
				type: NodeType.FILE,
				attributes: { rloc: 100, functions: 10, mcc: 1 },
				isExcluded: false,
				isFlattened: false,
				link: "http://www.google.de"
			},
			{
				name: "bigLeaf.java",
				path: "/root/bigLeaf.java",
				type: NodeType.FILE,
				attributes: { rloc: 100, functions: 10, mcc: 1 },
				isExcluded: false,
				isFlattened: false,
				link: "http://www.google.de"
			},
			{
				name: "Parent Leaf",
				path: "/root/Parent Leaf",
				type: NodeType.FOLDER,
				attributes: {},
				isExcluded: false,
				isFlattened: false,
				children: [
					{
						name: "smallLeaf.ts",
						path: "/root/Parent Leaf/smallLeaf.ts",
						type: NodeType.FILE,
						attributes: { rloc: 30, functions: 100, mcc: 100 },
						isExcluded: false,
						isFlattened: false
					},
					{
						name: "otherLeaf.ts",
						path: "/root/Parent Leaf/otherLeaf.ts",
						type: NodeType.FILE,
						attributes: { rloc: 70, functions: 1000, mcc: 10 },
						isExcluded: false,
						isFlattened: false
					},
					{
						name: "smallLeaf.java",
						path: "/root/Parent Leaf/smallLeaf.java",
						type: NodeType.FILE,
						attributes: { rloc: 30, functions: 100, mcc: 100 },
						isExcluded: false,
						isFlattened: false
					},
					{
						name: "otherLeaf.java",
						path: "/root/Parent Leaf/otherLeaf.java",
						type: NodeType.FILE,
						attributes: { rloc: 70, functions: 1000, mcc: 10 },
						isExcluded: false,
						isFlattened: false
					},
					{
						name: "anotherLeaf.java",
						path: "/root/Parent Leaf/anotherLeaf.java",
						type: NodeType.FILE,
						attributes: { rloc: 20, functions: 10, mcc: 2 },
						isExcluded: false,
						isFlattened: false
					}
				]
			}
		]
	}
})
