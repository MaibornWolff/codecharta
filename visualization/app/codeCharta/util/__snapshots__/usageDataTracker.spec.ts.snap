// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsageDataTracker trackEventUsageData should track node interaction blacklist event 1`] = `
"some-already-tracked-events-from-file-storage
{\\"eventType\\":\\"node_interaction\\",\\"eventTime\\":1612428357566,\\"payload\\":{\\"eventName\\":\\"ADD_BLACKLIST_ITEM\\",\\"id\\":\\"098f6bcd4621d373cade4e832627b4f6\\",\\"type\\":\\"exclude\\",\\"attributes\\":{}}}"
`;

exports[`UsageDataTracker trackEventUsageData should track node interaction focus event 1`] = `
"some-already-tracked-events-from-file-storage
{\\"eventType\\":\\"node_interaction\\",\\"eventTime\\":1612428357566,\\"payload\\":{\\"eventName\\":\\"FOCUS_NODE\\",\\"id\\":\\"9d1e46ce2260c90b8b7600a11ca08972\\"}}"
`;

exports[`UsageDataTracker trackEventUsageData should track setting changed event 1`] = `
"some-already-tracked-events-from-file-storage
{\\"eventType\\":\\"setting_changed\\",\\"eventTime\\":1612428357566,\\"payload\\":{\\"eventName\\":\\"SET_HEIGHT_METRIC\\",\\"newValue\\":\\"newHeightMetricValue\\"}}"
`;

exports[`UsageDataTracker trackMetaUsageData should track files with multiple programming languages properly 1`] = `"{\\"mapId\\":\\"invalid-md5-sample-checksum\\",\\"codeChartaApiVersion\\":\\"1.1\\",\\"creationTime\\":1612369999999,\\"exportedFileSizeInBytes\\":999,\\"statisticsPerLanguage\\":{\\"ts\\":{\\"metrics\\":{\\"rloc\\":{\\"avg\\":66.66666666666667,\\"max\\":100,\\"median\\":70,\\"min\\":30,\\"numberOfFiles\\":3,\\"totalSum\\":200},\\"functions\\":{\\"avg\\":370,\\"max\\":1000,\\"median\\":100,\\"min\\":10,\\"numberOfFiles\\":3,\\"totalSum\\":1110},\\"mcc\\":{\\"avg\\":37,\\"max\\":100,\\"median\\":10,\\"min\\":1,\\"numberOfFiles\\":3,\\"totalSum\\":111}},\\"numberOfFiles\\":3,\\"maxFilePathDepth\\":4,\\"avgFilePathDepth\\":3.6666666666666665},\\"java\\":{\\"metrics\\":{\\"rloc\\":{\\"avg\\":55,\\"max\\":100,\\"median\\":50,\\"min\\":20,\\"numberOfFiles\\":4,\\"totalSum\\":220},\\"functions\\":{\\"avg\\":280,\\"max\\":1000,\\"median\\":55,\\"min\\":10,\\"numberOfFiles\\":4,\\"totalSum\\":1120},\\"mcc\\":{\\"avg\\":28.25,\\"max\\":100,\\"median\\":6,\\"min\\":1,\\"numberOfFiles\\":4,\\"totalSum\\":113}},\\"numberOfFiles\\":4,\\"maxFilePathDepth\\":4,\\"avgFilePathDepth\\":3.75}}}"`;
